File: cnchi/ui/html/pages/_02_welcome/welcome.py
#!/usr/bin/env python
#  -*- coding: utf-8 -*-
#
#  welcome.py
#
#  Copyright © 2016 Antergos
#
#  This file is part of The Antergos Build Server, (AntBS).
#
#  AntBS is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License, or
#  (at your option) any later version.
#
#  AntBS is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  The following additional terms are in effect as per Section 7 of the license:
#
#  The preservation of all legal notices and author attributions in
#  the material or in the Appropriate Legal Notices displayed
#  by works containing it is required.
#
#  You should have received a copy of the GNU General Public License
#  along with AntBS; If not, see <http://www.gnu.org/licenses/>.


# Standard Lib
from _base_object import (
    json
)
from misc.extra import has_connection
from modules.update import UpdateModule
from ui.html.pages.html_page import HTMLPage, bg_thread


class WelcomePage(HTMLPage):
    """
    This page provides the user with a choice: (Try It or Install It).

    Class Attributes:
        Also see `HTMLPage.__doc__`

    """

    def __init__(self, name='welcome', *args, **kwargs):
        """
        Attributes:
            Also see `HTMLPage.__doc__`.

        Args:
            name (str): A name for this widget.

        """

        super().__init__(name=name, *args, **kwargs)

        self.signals.extend(['try-it-selected', 'install-it-selected', 'update-check',
                             '--update-available', 'restart', 'connection-check'])

        self._create_and_connect_signals()

    def _get_default_template_vars(self):
        signals = json.dumps(self.signals)
        return {'page_name': self.name, 'signals': signals}

    @bg_thread
    def connection_check_cb(self, *args):
        self._controller.trigger_js_event('connection-check-result', has_connection())

    def install_it_selected_cb(self, *args):
        self.go_to_next_page()

    def prepare(self):
        """ Prepare to become the current (visible) page. """
        pass

    def restart_cb(self, *args):
        self._controller.do_restart()

    def store_values(self):
        super().store_values()
        self._main_window.toggle_maximize()

    def try_it_selected_cb(self, *args):
        self._controller.exit_app()

    @bg_thread
    def update_check_cb(self, *args):
        updater = UpdateModule()

        for response in updater.do_update_check():
            self.logger.debug(response)
            if isinstance(response, str) and response.startswith('--'):
                self._controller.trigger_js_event(response)

            elif isinstance(response, dict):
                self._controller.trigger_js_event('update-check-result', response)

File: cnchi/ui/html/pages/_02_welcome/welcome.js
/*
 * welcome.js
 *
 * Copyright 2016 Antergos
 *
 * This file is part of Cnchi.
 *
 * Cnchi is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License,
 * or any later version.
 *
 * Cnchi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * The following additional terms are in effect as per Section 7 of the license:
 *
 * The preservation of all legal notices and author attributions in
 * the material or in the Appropriate Legal Notices displayed
 * by works containing it is required.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


class WelcomePage extends CnchiPage {

	constructor( id ) {
		if ( null !== window._page ) {
			return window._page;
		}

		super( id );

		window._page = this;
		this.signals = JSON.parse('{{ signals }}');
		this.$update_wrapper = $('.update_check_wrapper');
		this.$welcome_wrapper = $('.welcome_wrapper');
		this.$connection_wrapper = $('.no_internet_wrapper');
		this.has_connection = false;

		this.register_allowed_signals();
		// this.maybe_skip_to_next_page();
		this.register_event_handlers();
		this.initialize();
	}

	connection_check_result_cb( event, result ) {
		_page.has_connection = result;

		if ( false === result ) {
			_page.$connection_wrapper.fadeIn();
		}
	}

	do_has_connection_check() {
		if ( false === this.has_connection ) {
			cnchi.emit_signal('do-connection-check');
		}
	}

	do_update_check() {
		if ( false === this.has_connection ) {
			this.logger('An internet connection is required in order to check for installer update.');
			return;
		}

		if ( this.$connection_wrapper.is(':visible') ) {
			this.$connection_wrapper.fadeOut()
			.promise()
			.done(_show_update_wrapper);
		} else {
			_show_update_wrapper();
		}

		function _show_update_wrapper() {
			_page.$update_wrapper.fadeIn()
			.promise()
			.done(() => {
				cnchi.emit_signal('do-update-check');
			});
		}
	}

	initialize() {
		this.$welcome_wrapper.animateCss('magictime swashIn', () => {
			_page.$welcome_wrapper.addClass('showme');
		});
	}

	install_it_button_clicked_cb( event ) {
		console.log('clicked!');
		_page.$welcome_wrapper.animateCss('magictime swashOut', () => {
			console.log('callback!');
			let check, timer = 0;

			_page.$welcome_wrapper.hide(0);

			check = setInterval(() => {
				// Tell backend to do connection check every 10 seconds.
				if ( 0 === timer || timer % 10000 === 0 ) {
					_page.do_has_connection_check();
				}

				if ( true === _page.has_connection ) {
					clearInterval(check);
					_page.do_update_check();
				}

				timer += 500

			}, 250);
		});
	}

	maybe_skip_to_next_page() {
		let update_check_passed = localStorage.getItem('cnchi::welcome::update-check-passed');

		if ( null !== update_check_passed ) {
			cnchi.emit_signal('do-go-to-next-page');
		}
	}

	register_event_handlers() {
		$('.try_it button').on('click', this.try_it_button_clicked_cb);
		$('.install_it button').on('click', this.install_it_button_clicked_cb);
		$(window).on('update-available', this.update_available_cb);
		$(window).on('update-check-result', this.update_check_result_cb);
		$(window).on('connection-check-result', this.connection_check_result_cb)
	}

	try_it_button_clicked_cb( event ) {
		cnchi.emit_signal('do-try-it-selected');
	}

	update_available_cb() {
		let $heading = _page.$update_wrapper.find('h3');

		$heading.fadeOut()
		.promise()
		.done(() => {
			$heading.text("{{ _('An installer update is available. Please wait while the update is applied.') }}");
			$heading.fadeIn()
			.promise()
			.done(() => {
				$('.progress.active, .preloader-wrapper').toggleClass('active');
			});
		});
	}

	update_check_result_cb( event, data ) {
		let $headings = _page.$update_wrapper.find('h1, h3');

		$('.progress.active, .preloader-wrapper.active').toggleClass('hideme active');

		if ( true === data.result ) {
			localStorage.setItem('cnchi::welcome::update-check-passed', true);
		}

		if ( true === data.result && true === data.restart ) {
			// Update was installed successfully.
			$headings.fadeOut()
			.promise()
			.done(() => {
				$headings.filter('h1').text("{{ _('Update Complete.') }}");
				$headings.filter('h3').text("{{ _('Restarting the installer. One moment..') }}");
				$headings.fadeIn()
				.promise()
				.done(() => {
					setTimeout(() => {
						cnchi.emit_signal('do-restart');
					}, 4500);
				});
			});

		} else if ( true === data.result ) {
			// We already have the latest version.
			$headings.fadeOut()
			.promise()
			.done(() => {
				$headings.filter('h1').text("{{ _('Update Check Complete') }}");
				$headings.filter('h3').text("{{ _('You are using the latest version and can proceed with your installation!') }}");
				$headings.fadeIn()
				.promise()
				.done(() => {
					setTimeout(() => {
						_page.$update_wrapper.hide();
						cnchi.emit_signal('do-install-it-selected');
					}, 4500);
				});
			});

		} else {
			// Either the update check or the update itself failed.
		}
	}
}